### PR TITLE
IGNITE-12935 Disadvantages in log of historical rebalance

### PR DESCRIPTION
**Pull request by vldpyatkov**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java [hunk 1]
```diff
@@ -1750,10 +1751,18 @@ private boolean safeToUpdatePageMemories() {
 
         Map</*grpId*/Integer, Map</*partId*/Integer, /*updCntr*/Long>> grpPartsWithCnts = new HashMap<>();
 
-        for (Map.Entry<Integer, Map<Integer, CheckpointEntry>> e : earliestValidCheckpoints.entrySet()) {
+        for (Map.Entry<Integer, T2</*reason*/ReservationReason, Map</*partId*/Integer, CheckpointEntry>>> e : earliestValidCheckpoints.entrySet()) {
             int grpId = e.getKey();
 
-            for (Map.Entry<Integer, CheckpointEntry> e0 : e.getValue().entrySet()) {
+            if (e.getValue().get2() == null) {
```
Comment (ascherbakoff) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java:1757]: Looks like it's already logged in printReservationToLog, why we need to log it again in debug ?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java [hunk 2]
```diff
@@ -1772,9 +1781,60 @@ private boolean safeToUpdatePageMemories() {
...
+                    .map(entry -> '[' +
+                        entry.getValue().stream().map(grpId -> "[grpId=" + grpId +
+                            ", grpName=" + cctx.cache().cacheGroup(grpId).cacheOrGroupName() + ']')
+                            .collect(Collectors.joining(", ")) +
+                        ", reason=" + entry.getKey() + ']')
+                    .collect(Collectors.joining(", ")) + ']');
+        }
+
+        if (!F.isEmpty(reservedCachesToPrint)) {
+            log.info("Reserved cache groups with first reserved checkpoint IDs and reasons why previous checkpoint was inapplicable: [" +
```
Comment (ascherbakoff) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java:1826]: Reserved cache groups with first reserved checkpoint IDs and reasons why previous checkpoint was inapplicable
->
Cache groups with earliest reserved checkpoint and a reason why a previous checkpoint was inapplicable

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java [hunk 3]
```diff
@@ -1772,9 +1781,60 @@ private boolean safeToUpdatePageMemories() {
...
+            }
+            else {
+                reservedCachesToPrint.computeIfAbsent(entry.getValue().get1(), reason -> new ArrayList<>())
+                    .add(new T2(entry.getKey(), entry.getValue().get2().values().stream().min((cp1, cp2) ->
+                        Long.compare(cp1.timestamp(), cp2.timestamp())).get()));
+            }
+        }
+
+        if (!F.isEmpty(notReservedCachesToPrint)) {
+            log.info("Following caches were not reserved [" +
```
Comment (ascherbakoff) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java:1815]: Following caches were not reserved -> Caches groups were not reserved

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java [hunk 4]
```diff
@@ -4076,9 +4101,60 @@ else if (resetOwners)
             throw new IgniteException("Failed to assign partition states", e);
         }
 
+        if (log.isInfoEnabled() && !F.isEmpty(supplyInfoMap))
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java:4104]: Yes, but this leads to method invocation and other two after.
I think F.isEmpty(collection) is a common using pattern in our code.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java [hunk 5]
```diff
@@ -4076,9 +4101,60 @@ else if (resetOwners)
...
+    }
+
+    /**
+     * Does information contain partitions which will print to log.
+     *
+     * @param supplayInfoMap Map contains information about supplying partitions.
+     * @param reserved Reservation flag.
+     * @return True if map has partitions with same reserved flag, false otherwise.
+     */
+    private boolean hasPartitonToLog(Map<String, List<SupplyPartitionInfo>> supplayInfoMap, boolean reserved) {
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java:4147]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java (hunk 1 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java:1757]: Yes, you are right. I am removing it from here.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java (hunk 2 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java:1826]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java (hunk 3 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java:1815]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java [hunk 6]
```diff
@@ -3457,7 +3463,19 @@ private void assignHistoricalSuppliers(
                         deepestReserved.set(e0.getKey(), histCntr);
                 }
             }
+
+            //No one reservation matched for this partition.
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java:3467]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java [hunk 7]
```diff
@@ -4076,9 +4101,60 @@ else if (resetOwners)
...
         timeBag.finishGlobalStage("Assign partitions states");
     }
 
+    /**
+     * Prints detail information about partitions which did not have reservation
+     * history enough for historical rebalance.
+     *
+     * @param supplyInfoMap Map contains information about supplying partitions.
+     */
+    private void printPartitionRebalancingFully(Map<String, List<SupplyPartitionInfo>> supplyInfoMap) {
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java:4116]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java [hunk 8]
```diff
@@ -1772,9 +1781,60 @@ private boolean safeToUpdatePageMemories() {
...
         return grpPartsWithCnts;
     }
 
+    /**
+     * Prints detail information about caches which were not reserved
+     * and reservation depth for the caches which have WAL history enough.
+     *
+     * @param earliestValidCheckpoints Map contains information about caches' reservation.
+     */
+    private void printReservationToLog(
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java:1796]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ascherbakoff) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java:4104]: I mean log.isInfoEnabled() is not needed here

[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java]: I did it.

**Pull request closed without merging**